### ISSUE TITLE
Finish Test coverage

### ISSUE THREAD

**Issue by nevedomski** [nevedomski/mistral_ner#35]
# Test Coverage Implementation Plan - 90%+ Coverage

## Phase: 1 - Code Quality Foundation (Week 1)

**Priority**: HIGH
**Dependencies**: [NER-1](https://linear.app/homelab-nevedomski/issue/NER-1/ruff-formatting-and-linting) (Ruff), [NER-2](https://linear.app/homelab-nevedomski/issue/NER-2/add-mypy) (MyPy)
**Timeline**: 2-3 days

## Objective

Fix existing unit tests and achieve 90%+ test coverage for all modules, establishing a robust testing foundation for the ML pipeline.

## Current State Analysis

* Existing test files: `test_config.py`, `test_data.py`, `test_utils.py`
* Missing test files: `test_model.py`, `test_training.py`, `test_evaluation.py`
* Current coverage: Unknown (tests not running)
* CI/CD: Test job exists but may be failing

#### Update pyproject.toml test configuration:

```toml
[tool.pytest.ini_options]
minversion = "7.0"
testpaths = ["tests"]
python_files = "test_*.py"
python_classes = "Test*"
python_functions = "test_*"
addopts = [
    "-v",
    "--strict-markers",
    "--tb=short",
    "--cov=src",
    "--cov=api",
    "--cov=scripts",
    "--cov-report=term-missing",
    "--cov-report=html",
    "--cov-report=xml",
    "--cov-fail-under=90",
]
markers = [
    "slow: marks tests as slow (deselect with '-m \"not slow\"')",
    "integration: marks tests as integration tests",
    "gpu: marks tests that require GPU",
]

[tool.coverage.run]
source = ["src", "api", "scripts"]
omit = [
    "*/tests/*",
    "*/notebooks/*",
    "*/__init__.py",
    "*/setup.py",
]

[tool.coverage.report]
exclude_lines = [
    "pragma: no cover",
    "def __repr__",
    "raise AssertionError",
    "raise NotImplementedError",
    "if __name__ == .__main__.:",
    "if TYPE_CHECKING:",
]
```

### 2\. Create Test Fixtures ([conftest.py](http://conftest.py))

```python
# tests/conftest.py
import pytest
import tempfile
import shutil
from pathlib import Path
from unittest.mock import Mock, patch
import torch
from transformers import AutoTokenizer

@pytest.fixture
def temp_dir():
    """Create a temporary directory for test files."""
    temp_dir = tempfile.mkdtemp()
    yield Path(temp_dir)
    shutil.rmtree(temp_dir)

@pytest.fixture
def mock_tokenizer():
    """Mock tokenizer for testing."""
    tokenizer = Mock()
    tokenizer.pad_token_id = 0
    tokenizer.model_max_length = 256
    tokenizer.return_value = {
        "input_ids": [[101, 2023, 2003, 102]],
        "attention_mask": [[1, 1, 1, 1]]
    }
    return tokenizer

@pytest.fixture
def mock_model():
    """Mock model for testing."""
    model = Mock()
    model.config.num_labels = 9
    model.device = torch.device("cpu")
    return model

@pytest.fixture
def sample_config():
    """Sample configuration for testing."""
    from src.config import Config
    config = Config()
    config.training.num_train_epochs = 1
    config.training.per_device_train_batch_size = 2
    return config

@pytest.fixture
def sample_dataset():
    """Sample NER dataset for testing."""
    return {
        "train": [
            {"tokens": ["John", "lives", "in", "New", "York"], 
             "ner_tags": [1, 0, 0, 3, 4]},
            {"tokens": ["Apple", "Inc", "is", "great"], 
             "ner_tags": [5, 6, 0, 0]}
        ]
    }
```

#### 3.1 test_model.py (NEW)

```python
"""Tests for model module."""
import pytest
from unittest.mock import patch, Mock
import torch
from src.model import setup_model, save_model_checkpoint, load_model_from_checkpoint

class TestModelSetup:
    """Test model setup functions."""
    
    @patch('src.model.AutoModelForTokenClassification.from_pretrained')
    @patch('src.model.AutoTokenizer.from_pretrained')
    def test_setup_model_8bit(self, mock_tokenizer, mock_model, sample_config):
        """Test 8-bit model setup."""
        # Test implementation
        
    def test_setup_model_cpu(self, sample_config):
        """Test CPU model setup."""
        # Test implementation
        
    @pytest.mark.gpu
    def test_setup_model_gpu(self, sample_config):
        """Test GPU model setup."""
        # Test implementation

class TestModelCheckpointing:
    """Test model checkpointing."""
    
    def test_save_model_checkpoint(self, mock_model, mock_tokenizer, temp_dir):
        """Test saving model checkpoint."""
        # Test implementation
        
    def test_load_model_from_checkpoint(self, temp_dir, sample_config):
        """Test loading model from checkpoint."""
        # Test implementation
```

#### 3.2 test_training.py (NEW)

```python
"""Tests for training module."""
import pytest
from unittest.mock import patch, Mock
from src.training import run_training_pipeline, MemoryCallback, ClearCacheCallback

class TestTrainingPipeline:
    """Test training pipeline."""
    
    @patch('src.training.Trainer')
    def test_run_training_pipeline(self, mock_trainer, mock_model, sample_config):
        """Test training pipeline execution."""
        # Test implementation
        
    def test_memory_callback(self):
        """Test memory monitoring callback."""
        # Test implementation
        
    @pytest.mark.slow
    def test_full_training_cycle(self, sample_dataset, sample_config):
        """Integration test for full training cycle."""
        # Test implementation
```

#### 3.3 test_evaluation.py (NEW)

```python
"""Tests for evaluation module."""
import pytest
import numpy as np
from src.evaluation import compute_metrics, evaluate_model, print_evaluation_report

class TestEvaluationMetrics:
    """Test evaluation metrics."""
    
    def test_compute_metrics_perfect(self):
        """Test metrics computation with perfect predictions."""
        predictions = np.array([[1, 2, 3], [4, 5, 6]])
        labels = np.array([[1, 2, 3], [4, 5, 6]])
        # Test implementation
        
    def test_compute_metrics_partial(self):
        """Test metrics with partial matches."""
        # Test implementation
        
    def test_evaluate_model(self, mock_model, sample_dataset):
        """Test model evaluation."""
        # Test implementation
```

#### 3.4 Enhanced existing tests

* **test_config.py**: Add edge cases, error handling
* **test_data.py**: Add tokenization edge cases, label alignment
* **test_utils.py**: Add GPU memory tests, logging tests

### 4\. Test Categories and Coverage Targets

| Module | Current Coverage | Target Coverage | Priority |
| -- | -- | -- | -- |
| [config.py](http://config.py) | \~60% | 95% | High |
| [data.py](http://data.py) | \~40% | 90% | High |
| [model.py](http://model.py) | 0% | 85% | Critical |
| [training.py](http://training.py) | 0% | 85% | Critical |
| [evaluation.py](http://evaluation.py) | 0% | 90% | High |
| [utils.py](http://utils.py) | \~30% | 95% | Medium |
| api/serve.py | 0% | 90% | Medium |
| scripts/\*.py | 0% | 80% | Low |

### 5\. Testing Best Practices for ML

1. **Mock Heavy Operations**
   * Model loading/saving
   * Training loops
   * GPU operations
2. **Parametrize Tests**

   ```python
   @pytest.mark.parametrize("batch_size,seq_length", [
       (1, 128), (4, 256), (16, 512)
   ])
   def test_data_collator(batch_size, seq_length):
       # Test implementation
   ```
3. **Test Edge Cases**
   * Empty inputs
   * Single token sequences
   * Maximum length sequences
   * Special characters
4. **Integration Tests**
   * End-to-end pipeline (marked as slow)
   * API endpoint testing
   * Model inference testing

### 6\. CI/CD Updates

Update `.github/workflows/ci.yml`:

```yaml
      - name: Run tests with coverage
        run: |
          pytest tests/ -v --cov=src --cov=api --cov=scripts \
            --cov-report=xml --cov-report=term-missing \
            --cov-fail-under=90
          
      - name: Generate coverage report
        run: |
          coverage html
          
      - name: Upload coverage reports
        uses: actions/upload-artifact@v4
        with:
          name: coverage-report
          path: htmlcov/
```

## Validation Criteria

- [ ] All existing tests pass
- [ ] 90%+ coverage for src/ modules
- [ ] 90%+ coverage for api/ module
- [ ] 80%+ coverage for scripts/
- [ ] CI/CD test job passes
- [ ] Coverage reports generated

## Benefits

1. **Reliability**: Catch bugs before production
2. **Refactoring Safety**: Confidence in code changes
3. **Documentation**: Tests serve as usage examples
4. **Quality Gate**: Enforce standards via CI/CD

## Timeline

1. **Day 1**: Fix infrastructure, update existing tests
2. **Day 2**: Create new test files for missing modules
3. **Day 3**: Achieve coverage targets, update CI/CD
